Documentation about gutter icons (#2385)
dafny-lang/dafny: docs/DafnyRef/UserGuide.md

    Emacs](https://github.com/boogie-org/boogie-friends).

- VSCode, a cross-platform editor for many programming languages has an extension for dafny,
  installed from within VSCode. VSCode is available [here](http://code.visualstudio.com).
  [More information about this extension](#sec-dafny-language-server-vscode)
  The extension provides syntax highlighting, in-line parser,
  type and verification errors, code navigation, counter-example display and gutter highlights.
  

- An old Visual Studio plugin is no longer supported

Information about installing IDE extensions for Dafny is found
on the [Dafny INSTALL page in the wiki](https://github.com/dafny-lang/dafny/wiki/INSTALL).

## 24.6. Dafny VSCode extension and the Dafny Language Server {#sec-dafny-language-server-vscode}

Dafny [implements](https://github.com/dafny-lang/dafny/tree/master/Source/DafnyLanguageServer) the official [Language Server Protocol](https://microsoft.github.io/language-server-protocol/).
This language server is used within the Dafny VSCode Extension, and it currently offers the following features:
- Quick syntax highlighting
- Parsing, Resolution and realtime verification diagnostics

> - Support for Dafny plugins

Author: ```suggestion
- Support for [Dafny plugins](https://github.com/dafny-lang/dafny/tree/master/Source/DafnyLanguageServer#plugins)
```
```suggestion
- Support for Dafny plugins
```

- Moderate support for symbol completion
- Moderate support for code navigation
- Counter-example display
- Ghost statements highlighting
- Gutter highlights.

### 24.6.1. Gutter highlights meaning {#sec-gutter-highlights}

The first time a file is loaded, the gutter will highlight in a transparent squiggly green line all the methods that need to be verified, like this:

![image](https://user-images.githubusercontent.com/3601079/178058374-a1e5e9ca-2c5e-493a-bb60-d00310962741.png)

When the file is saved (in verification on save), or whenever the Dafny verifier is ready (in verification on change), it will start to verify methods.

> The gutter will turn as a thin green rectangle on methods that have been verified, and display an animated less transparent green squiggly line on methods that are being actively verified:

Reviewer: ```suggestion
That line will turn into a thin green rectangle on methods that have been verified, and display an animated less transparent green squiggly line on methods that are being actively verified:
```

![image](https://user-images.githubusercontent.com/3601079/178058713-1266a23f-fdb4-4494-844a-488cc214c797.png)

When the verification finishes, if a method, a function, a constant with default initialization or a subset type with a witness has some verification errors in it,

> Dafny will display two yellow vertical rails indicating an error context.

Reviewer: ```suggestion
the editor will display two yellow vertical rails indicating an error context.
```

![image](https://user-images.githubusercontent.com/3601079/178058908-71425938-57a4-454d-805d-2923d7c4de73.png)

Inside this context, if there is a failing assertion on a line,
it will fill the gap between the vertical yellow bars with a red rectangle, even if there might be other assertions that are verified on the line.
If there is no error on a line, but there is at least one assertion that verified, it will display a green disk with a white checkmark on it,

> which can be used to indicate progress in a proof.

Reviewer: ```suggestion
which can be used to check progress in a proof search.
```

> As soon as a line is changed, the gutter icons turn transparent and squiggly, to indicate they are obsolete.

Reviewer: ```suggestion
As soon as a line is changed, the gutter icons turn transparent and squiggly, to indicate their obsolescence.
```

![image](https://user-images.githubusercontent.com/3601079/178063330-dd346ddf-aa60-487e-a368-3a4af572aeac.png)

> The red error rectangles have half the size, so that visually it's clear that they might be obsolete.

Reviewer: ```suggestion
The red error rectangles occupy only half the horizontal space, to visualise their possible obsolescence.
```

When the file is saved (in verification on save), or as soon as possible otherwise,

> these squiggly icons will be animated whenever the Dafny verifier inspect the area.

Reviewer: ```suggestion
these squiggly icons will be animated while the Dafny verifier inspect the area.
```

![image](https://user-images.githubusercontent.com/3601079/178063848-d0fb02d9-c743-4ffc-9e34-c5f8731c79d1.png)

If the method was verifying before a change, instead of two yellow vertical bars with a red squiggly line,
the gutter icons display an animated squiggly but more firm green line, thereby indicating that the method used to verify,
but Dafny is still re-verifying it.

![image](https://user-images.githubusercontent.com/3601079/178068281-f15fa83f-6180-4de1-ab63-b65957f6c86b.png)

> If there is a parse or resolution error, the previous gutter icons turn gray and a red triangle indicate 
> the line where the parse or resolution error is.

Reviewer: ```suggestion
If there is a parse or resolution error, the previous gutter icons turn gray and a red triangle indicates 
the position of the parse or resolution error.
```
